plugins/allocators/bitmap/bitmap_ipv4.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2018-present the CoreDHCP Authors. All rights reserved
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package bitmap
+
+// This allocator handles IPv4 assignments with a similar logic to the base bitmap, but a simpler
+// implementation due to the ability to just use uint32 for IPv4 addresses
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/coredhcp/coredhcp/plugins/allocators"
+	"github.com/willf/bitset"
+)
+
+var (
+	// ErrNoAddrAvail is returned when we can't allocate an IP
+	// because there's no unallocated space left
+	ErrNoAddrAvail = errors.New("No address available to allocate")
+
+	errNotInRange = errors.New("IP outside of allowed range")
+
+	errInvalidIP = errors.New("Invalid IP passed as input")
+)
+
+// IPv4Allocator allocates IPv4 addresses, tracking utilization with a bitmap
+type IPv4Allocator struct {
+	start uint32
+	end   uint32
+
+	bitmap *bitset.BitSet
+	l      sync.Mutex
+}
+
+func (a *IPv4Allocator) toIP(offset uint32) net.IP {
+	if offset > a.end-a.start {
+		return nil
+	}
+
+	r := make(net.IP, 4)
+	binary.BigEndian.PutUint32(r, a.start+offset)
+	return r
+}
+
+func (a *IPv4Allocator) toOffset(ip net.IP) (uint, error) {
+	if ip.To4() == nil {
+		return 0, errInvalidIP
+	}
+
+	intIP := binary.BigEndian.Uint32(ip.To4())
+	if intIP < a.start || intIP > a.end {
+		return 0, errNotInRange
+	}
+
+	return uint(intIP - a.start), nil
+}
+
+// Allocate reserves an IP for a client
```

Comment:
can you add a comment on how the hint works for the user?

Reply:
This is just the implementation, there's a user-level explanation in the [interface](https://github.com/coredhcp/coredhcp/blob/f4b74e59cadf52f325dc856d7b3fa6295bb112cc/plugins/allocators/allocator.go#L21)